docs/src/lang/option-types.md

```diff
@@ -0,0 +1,386 @@
+
+**Arguments:**
+
+- `o` an optional value
+
+**Effect:** `OptionToSeq(o)` is `<<v>>` iff `o = Some(v)`, or else `<<>>`.
+
+**Determinism:** Deterministic.
+
+**Errors:** No errors.
+
+**Example in TLA+:**
+
+```tla
+LET \* @type: Seq(Int);
+    empty == <<>>
+IN
+/\ OptionToSeq(None) = empty
+/\ OptionToSeq(Some(1)) = <<1>>
+```
+
+#### Converting to sets
+
+**Notation:** `OptionToSet(o)`
+
+**LaTeX notation:** same
+
+**Apalache type:** `(Some(a) | None(UNIT)) => Seq(a)`, for some type `a`.
+
+**Arguments:**
+
+- `o` an optional value
+
+**Effect:** `OptionToSet(o)` is like `OptionToSeq`, but producing a set.
+
+**Determinism:** Deterministic.
+
+**Errors:** No errors.
+
+**Example in TLA+:**
+
+```tla
+LET \* @type: Set(Int);
+    empty == {}
+IN
+/\ OptionToSet(None) = empty
+/\ OptionToSet(Some(1)) = {1}
+```
+
+----------------------------------------------------------------------------
+
+### Obtaining an optional value from a set
+
+**Notation:** `OptionGuess(s)`
+
+**LaTeX notation:** same
+
+**Apalache type:** `Set(a) => Some(a) | None(UNIT)`, for some type `a`.
+
+**Arguments:**
+
+- `s` is a set
+
+**Effect:** `OptionGuess(s)` is `None` if `s = {}`, otherwise it is `Some(x)`,
+where `x \in s`. `x` is selected from `s` nondeterministically.
+
+**Determinism:** Nondeterministic.
+
+**Errors:** No errors.
+
+**Example in TLA+:**
+
+```tla
+LET
+    \* @type: Set(Int);
+    empty == {}
+IN
+/\ OptionGuess(empty) = None
+/\ LET choices == {1,2,3,4} IN
+    LET choice == OptionGuess(choices) IN
+    VariantGetUnsafe("Some", choice) \in choices
+```
+
+----------------------------------------------------------------------------
+
+<a name="OptionFunApp"></a>
+### Apply a function to a partial value
+
+**Notation:** `OptionFunApp(f, o)`
+
+**LaTeX notation:** same
+
+**Apalache type:** `(a -> b, Some(a) | None(UNIT)) => Some(b) | None(UNIT)`
+
+**Arguments:**
+
+- `f` is a function
+- `o` is an optional value
+
+**Effect:** `OptionFunApp(f, o)` is `Some(f[v])` if `o = Some(v)` or else `None`.
+
+**Determinism:** Deterministic.
+
+**Errors:** No errors.
+
+**Example in TLA+:**
+
+```tla
+LET f == [x \in 1..3 |-> x + 1] IN
+/\ OptionFunApp(f, Some(1)) = Some(2)
+/\ OptionFunApp(f, None) = None
+```
+
+----------------------------------------------------------------------------
+
+<a name="OptionPartialFun"></a>
+### Extend a total function into a partial function
+
+**Notation:** `OptionPartialFun(f, undef)`
+
+**LaTeX notation:** same
+
+**Apalache type:** `(a -> b, Set(a)) => (a -> Some(b) | None(UNIT))`
+
+**Arguments:**
+
+- `f` is a total function
+- `undef` is a set of values for which the new function is to be "undefined"
+
+**Effect:** `OptionPartialFun(f, undef)` is a function mapping each value in
+`undef` to `None`, and each value `x \in (DOMAIN f \ undef)` to `Some(f[x])`.
+This can be used to extend a total function a "partial function" whose domain is
+expanded to include the vaules in 'undef'.
```

Comment:
```suggestion
This can be used to extend a total function into a "partial function" whose domain is
extended to include the values in 'undef'.
```

Reply:
oop. swear I thought I proof read these :facepalm: Thanks for the fixes here!